File: apps/android/app/src/testQa/java/chat/rogi/rogichat/preview/WireframeState.kt
package chat.rogi.rogichat.preview

import chat.rogi.rogichat.core.navigation.*
import chat.rogi.rogichat.feature.settings.*

// UI-only state. This type must never become a session, repository, or API DTO.
enum class PreviewRole(val label: String) { FAN("팬"), STREAMER("스트리머") }
typealias PreviewPage = AppPage
enum class ListScenario(val label: String) { CONTENT("목록"), LOADING("로딩"), EMPTY("빈 목록"), ERROR("오류") }
enum class PreviewAudience(val label: String) { PRIVATE("개인 답장"), SHARED("전체 대화") }
data class SampleRoom(val id: String, val title: String, val summary: String)

object WireframeFixtures {
    val rooms = listOf(
        SampleRoom("sample-room-a", "샘플 스트리머 A", "함께 나누는 오늘의 이야기"),
        SampleRoom("sample-room-b", "샘플 스트리머 B", "다음 대화를 기다리고 있어요"),
    )
    val fans = listOf("샘플 팬 A", "샘플 팬 B")
}

data class WireframeState(
    val role: PreviewRole = PreviewRole.FAN,
    val navigation: ShellNavigation = ShellNavigation(),
    val roomId: String? = null,
    val singleRoomMode: Boolean = false,
    val scenario: ListScenario = ListScenario.CONTENT,
    val audience: PreviewAudience = PreviewAudience.PRIVATE,
    val target: String? = null,
    val draft: String = "",
    val profile: ProfileEditor = ProfileEditor("샘플 프로필"),
) {
    val page: AppPage get() = navigation.page
    val linkPreviewPassed: Boolean get() = navigation.access == ShellAccess.READY
    fun switchAccess(value: ShellAccess) = WireframeState(role = role, navigation = navigation.withAccess(value))
    fun selectTab(value: AppTab) = copy(navigation = navigation.selectTab(value))
    fun switchRole(value: PreviewRole) = if (value == role) this else WireframeState(role = value)
    fun previewLink() = if (page == PreviewPage.WELCOME) switchAccess(ShellAccess.LINK_REQUIRED) else this
    val visibleRooms: List<SampleRoom> get() = if (singleRoomMode) WireframeFixtures.rooms.take(1) else WireframeFixtures.rooms
    fun previewRooms(singleRoom: Boolean = false): WireframeState {
        if (page != PreviewPage.LINK) return this
        val next = switchAccess(ShellAccess.READY).copy(singleRoomMode = singleRoom)
        return if (singleRoom) next.openRoom(next.visibleRooms.first().id) else next
    }
    fun openRoom(id: String): WireframeState {
        if (!linkPreviewPassed || page != PreviewPage.ROOMS || scenario != ListScenario.CONTENT ||
            visibleRooms.none { it.id == id }) return this
        return copy(navigation = navigation.open(AppPage.CHAT), roomId = id, draft = "", target = null, audience = PreviewAudience.PRIVATE)
    }
    fun open(value: PreviewPage) = copy(navigation = navigation.open(value))
    fun changeAudience(value: PreviewAudience): WireframeState {
        if (role != PreviewRole.STREAMER || page != PreviewPage.CHAT) return this
        // A draft must never silently move to a different recipient/audience.
        return if (value == audience) this else copy(audience = value, target = null, draft = "")
    }
    fun selectTarget(value: String): WireframeState {
        if (role != PreviewRole.STREAMER || page != PreviewPage.CHAT || value !in WireframeFixtures.fans) return this
        return if (target == value && audience == PreviewAudience.PRIVATE) this
        else copy(audience = PreviewAudience.PRIVATE, target = value, draft = "")
    }
    val canCompose: Boolean get() = page == PreviewPage.CHAT && roomId != null &&
        (role == PreviewRole.FAN || audience == PreviewAudience.SHARED || target != null)
    fun editProfile(value: String) = if (page == AppPage.PROFILE) copy(profile = profile.edit(value)) else this
    fun discardProfile() = if (page == AppPage.PROFILE) copy(profile = profile.discard()) else this
    fun profilePhase(value: ProfilePhase) = if (page == AppPage.PROFILE) copy(profile = profile.copy(phase = value)) else this
    fun editDraft(value: String) = if (canCompose) copy(draft = value.take(2000)) else this
    fun back(): WireframeState {
        val next = navigation.back()
        return if (AppPage.CHAT !in next.talkPath) copy(navigation = next, roomId = null, draft = "", target = null)
        else copy(navigation = next)
    }
}

File: apps/ios/Tests/Fixtures/WireframeState.swift
// UI-only state. Never use these values as a session, repository, or API DTO.
enum PreviewRole: String, CaseIterable {
    case fan = "팬", streamer = "스트리머"
}
typealias PreviewPage = AppPage
enum ListScenario: String, CaseIterable {
    case content = "목록", loading = "로딩", empty = "빈 목록", error = "오류"
}
enum PreviewAudience: String, CaseIterable {
    case `private` = "개인 답장", shared = "전체 대화"
}
struct SampleRoom: Identifiable {
    let id: String
    let title: String
    let summary: String
}
enum WireframeFixtures {
    static let rooms = [
        SampleRoom(id: "sample-room-a", title: "샘플 스트리머 A", summary: "함께 나누는 오늘의 이야기"),
        SampleRoom(id: "sample-room-b", title: "샘플 스트리머 B", summary: "다음 대화를 기다리고 있어요")
    ]
    static let fans = ["샘플 팬 A", "샘플 팬 B"]
}

struct WireframeState {
    private(set) var role: PreviewRole = .fan
    private(set) var navigation = ShellNavigation()
    var linkPreviewPassed: Bool { navigation.access == .ready }
    private(set) var singleRoomMode = false
    var visibleRooms: [SampleRoom] { singleRoomMode ? Array(WireframeFixtures.rooms.prefix(1)) : WireframeFixtures.rooms }
    private(set) var roomID: String?
    var scenario: ListScenario = .content
    private(set) var audience: PreviewAudience = .private
    private(set) var target: String?
    private(set) var draft = ""
    private(set) var profile = ProfileEditor(baseline: "샘플 프로필")

    var page: AppPage { navigation.page }
    var canCompose: Bool {
        page == .chat && roomID != nil && (role == .fan || audience == .shared || target != nil)
    }
    mutating func reset() { self = WireframeState(role: role) }
    mutating func switchRole(_ value: PreviewRole) {
        if value != role { self = WireframeState(role: value) }
    }
    mutating func switchAccess(_ value: ShellAccess) {
        self = WireframeState(role: role)
        navigation.setAccess(value)
    }
    mutating func selectTab(_ value: AppTab) { navigation.selectTab(value) }
    mutating func previewLink() {
        guard page == .welcome else { return }
        switchAccess(.linkRequired)
    }
    mutating func previewRooms(singleRoom: Bool = false) {
        guard page == .link else { return }
        switchAccess(.ready)
        singleRoomMode = singleRoom
        if singleRoom { openRoom(visibleRooms[0].id) }
    }
    mutating func openRoom(_ id: String) {
        guard linkPreviewPassed, page == .rooms, scenario == .content,
              visibleRooms.contains(where: { $0.id == id }) else { return }
        roomID = id
        draft = ""
        target = nil
        audience = .private
        navigation.open(.chat)
    }
    mutating func open(_ value: PreviewPage) { navigation.open(value) }
    mutating func changeAudience(_ value: PreviewAudience) {
        guard role == .streamer, page == .chat, audience != value else { return }
        audience = value
        target = nil
        draft = ""
    }
    mutating func selectTarget(_ value: String) {
        guard role == .streamer, page == .chat, WireframeFixtures.fans.contains(value) else { return }
        if target == value && audience == .private { return }
        audience = .private
        target = value
        draft = ""
    }
    mutating func editProfile(_ value: String) { if page == .profile { profile.edit(value) } }
    mutating func discardProfile() { if page == .profile { profile.discard() } }
    mutating func profilePhase(_ value: ProfilePhase) { if page == .profile { profile.phase = value } }
    mutating func editDraft(_ value: String) {
        guard canCompose else { return }
        draft = String(value.prefix(2000))
    }
    mutating func pop(to newPath: [AppPage], in tab: AppTab) {
        navigation.pop(to: newPath, in: tab)
        if !navigation.talkPath.contains(.chat) {
            roomID = nil
            target = nil
            draft = ""
        }
    }
}
